File: binary-search-tree.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>

template <typename T>
class Node {

private:
  T x;
  Node<T>* parent = nullptr;
  Node<T>* left = nullptr;
  Node<T>* right = nullptr;

public:
  Node<T> (T _x, Node<T>* p, Node<T>* l, Node<T>* r){
    x = _x; parent = p; left = l; right = r;
  }
  Node<T> (T _x, Node<T>* p){
    (*this) = Node<T>(_x, p, nullptr, nullptr);
  }
  Node<T> (T _x, T _p, T _l, T _r){
    (*this) = Node<T>(_x, new Node<T>(_p), new Node<T>(_l), new Node<T>(_r));
  }
  Node<T> (T _x) {x = _x;}

  T getX() {return x;}
  Node<T>* getParent() {return parent;}
  Node<T>* getLeft() {return left;}
  Node<T>* getRight() {return right;}

  void setX(T _x) {x = _x;}
  void setParent(Node<T>* p) {parent = p;}
  void setLeft(Node<T>* l) {left = l;}
  void setRight(Node<T>* r) {right = r;}
  void setParent(T _p) {parent = new Node<T>(_p);}
  void setLeft(T _l) {left = new Node<T>(_l);}
  void setRight(T _r) {right = new Node<T>(_r);}

  int nb_of_children(){
    return (left != nullptr) + (right != nullptr);
  }

  bool isLeaf(){
    return nb_of_children() == 0;
  }

  void print(){
    std::cout << "NODE: \n";
    std::cout << "|  x      = " << x << '\n';
    std::cout << "|  parent = " << (parent == nullptr ? -1 : parent->getX()) << '\n';
    std::cout << "|  left   = " << (left == nullptr ? -1 : left->getX()) << '\n';
    std::cout << "|  right  = " << (right == nullptr ? -1 : right->getX()) << '\n';
    std::cout << "------------\n";
  }
};

template <typename T>
class BST {

private:
  Node<T>* root = nullptr;
  int sz = 0;

  void add(Node<T>* u, Node<T>* p, T x, bool left){
    if (!u) {
      u = new Node<T>(x, p);
      if (left) p->setLeft(u); else p->setRight(u);
      return;
    }
    if (x < u->getX()) add(u->getLeft(), u, x, 1);
    else add(u->getRight(), u, x, 0);
  }

  int height(Node<T>* u){
    if (!u) return 0;
    return 1 + std::max(height(u->getLeft()), height(u->getRight()));
  }

  Node<T>* find(Node<T>* u, T x){
    if (!u) return u;
    if (x < u->getX()) return find(u->getLeft(), x);
    if (x > u->getX()) return find(u->getRight(), x);
    return u;
  }

  void dfs(Node<T>* u, std::vector<Node<T>*>& v){
    if (!u) return;
    v.push_back(u);
    dfs(u->getLeft()); dfs(u->getRight());
  }

  void print_indented(Node<T>* u, std::string x, int l){
    if (!u) return;
    std::cout << x << (l == 1 ? "lt_" : (l == 0 ? "rt_" : "")) << u->getX() << '\n';
    print_indented(u->getLeft(), x + "| ", 1);
    print_indented(u->getRight(), x + "| ", 0);
  }

public:
  BST() {}

  int size() {return sz;}
  bool isEmpty() {return size() == 0;}
  int height() {return height(getRoot()) - 1;}
  Node<T>* getRoot() {return root;}

  Node<T>* getMin(Node<T>* u){
    while (u && u->getLeft()) u = u->getLeft();
    return u;
  }

  void add(T x){
    if (!root) root = new Node<T>(x);
    else add(getRoot(), nullptr, x, 0);
  }

  Node<T>* find(T x) {return find(getRoot(), x);}

  void remove(T x){
    Node<T>* u = find(x);
    if (!u) return;
    --sz;
    if (u->isLeaf()){
      Node<T>* p = u->getParent();
      if (!p) root = nullptr;
      else {
        if (p->getLeft() == u) p->setLeft(nullptr);
        else p->setRight(nullptr);
      }
      return;
    }
    if (!u->getRight()){
      Node<T>* p = u->getParent();
      if (!p) root = nullptr;
      else {
        if (p->getLeft() == u) p->setLeft(u->getLeft());
        else p->setRight(u->getLeft());
      }
      return;
    }
    if (!u->getLeft()){
      Node<T>* p = u->getParent();
      if (!p) root = nullptr;
      else {
        if (p->getLeft() == u) p->setLeft(u->getRight());
        else p->setRight(u->getRight());
      }
      return;
    }
    ++sz;
    Node<T>* successor = getMin(u->getRight());
    const T k = successor->getX();
    remove(k); u->setX(k);
    return;
  }

  void print(Node<T>* u){
    if (!u) return;
    std::cout << u->getX() << ' ';
    print(u->getLeft()); print(u->getRight());
  }

  std::vector<Node<T>*> dfs(Node<T>* u) {
    std::vector<Node<T>*> v;
    dfs(u, v);
    return v;
  }

  std::vector<Node<T>*> bfs(Node<T>* u){
    std::vector<Node<T>*> arr;
    std::queue<Node<T>*> q;
    if (u) q.push(u);
    while (!q.empty()){
      Node<T>* v = q.front(); q.pop();
      arr.push_back(v);
      if (v->getLeft()) q.push(v->getLeft());
      if (v->getRight()) q.push(v->getRight());
    }
    return arr;
  }

  void print_indented(){
    print_indented(getRoot(), "", -1);
  }
};

int main(){
  BST<int> tree;
  for (int i = 1; i <= 3; ++i){
    tree.add(2 * i); tree.add(i); tree.add(2 * i + 1);
  }
  tree.print_indented();
}

File: binary-search-tree.java
import java.util.*;

class Node<T> {
  // ------------------- Data Members ------------------- //
  private T x;
  private Node<T> parent;
  private Node<T> left;
  private Node<T> right;
  
  // ------------------- Constructors ------------------- //
  public Node (T _x, Node<T> _parent, Node<T> _left, Node<T> _right){
    x = _x; parent = _parent; left = _left; right = _right;
  }
  public Node (T _x){
    this(_x, null, null, null);
  }
  public Node (T _x, Node<T> _parent){
    this(_x, _parent, null, null);
  }

  // ------------------- Getters and Setters ------------------- //
  public T getX() {return x;}
  public Node<T> getParent() {return parent;}
  public Node<T> getLeft() {return left;}
  public Node<T> getRight() {return right;}

  public void setX(T _x) {x = _x;}
  public void setParent(Node<T> _parent) {parent = _parent;}
  public void setLeft(Node<T> _left) {left = _left;}
  public void setRight(Node<T> _right) {right = _right;}

  // ------------------- Main Functions ------------------- //
  public int nbChildren() {
    return (left == null ? 0 : 1) + (right == null ? 0 : 1);
  }

  public boolean isRoot() {return parent == null;}
  public boolean isLeaf() {return nbChildren() == 0;}
}

class BinarySearchTree<T extends Comparable<? super T>> {
  private Node<T> root = null; // pointer to the node of the root
  private int sz = 0; // nb of nodes in the tree

  public BinarySearchTree() {}

  public int size() {return sz;}
  public Node<T> root() {return root;}

  public boolean isEmpty() {return sz == 0;}
  public int height() {return getHeight(root()) - 1;}

  //---------------- TREE EXTRA OPERATION ------------------//

  /**
   * return the node with the minimum value in the tree
   * i.e. the left-most node
   * @param curr node from where to start the search
   * @return the minimum node
   */
  public Node<T> getMinNode(Node<T> curr){
    while (curr != null && curr.getLeft() != null) curr = curr.getLeft();
    return curr;
  }

  /**
   * recursively computes the height of the tree
   * @param curr node from where to start the search
   * @return height of the tree
   */
  private int getHeight(Node<T> curr){
    if (curr == null) return 0;
    return 1 + Math.max(getHeight(curr.getLeft()), getHeight(curr.getRight()));
  }

  //---------------- TREE OPERATION ------------------//

  /**
   * Adds a new node to the BST with value x
   * @param x value of the new node
   */
  public void add(T x){
    Node<T> nw = new Node(x); ++sz;
    if (root == null) {root = nw; return;}
    Node<T> curr = root;
    while (true){
      if (x.compareTo(curr.getX()) < 0) {
        if (curr.getLeft() == null) {curr.setLeft(nw); return;}
        curr = curr.getLeft();
      }
      else {
        if (curr.getRight() == null) {curr.setRight(nw); return;}
        curr = curr.getRight();
      }
    }
  }

  /**
   * removes the first node that has the value x
   * @param x value of the node to remove
   * @return True if we found the node (removed it), and false otherwise
   */
  public boolean remove(T x){
    Node<T> curr = root();
    while (curr != null && x != curr.getX()){
      if (x.compareTo(curr.getX()) < 0) curr = curr.getLeft();
      else curr = curr.getRight();
    }
    if (curr == null) return false;
    if (curr.isLeaf()){
      Node<T> parent = curr.getParent();
      if (parent == null) root = null;
      else {
        if (parent.getLeft() == curr) parent.setLeft(null);
        else parent.setRight(null);
      }
      --sz; return true;
    }
    if (curr.getRight() == null){
      Node<T> parent = curr.getParent();
      if (parent == null) root = null;
      else {
        if (parent.getLeft() == curr) parent.setLeft(curr.getLeft());
        else parent.setRight(curr.getLeft());
      }
      --sz; return true;
    }
    if (curr.getLeft() == null){
      Node<T> parent = curr.getParent();
      if (parent == null) root = null;
      else {
        if (parent.getLeft() == curr) parent.setLeft(curr.getRight());
        else parent.setRight(curr.getRight());
      }
      --sz; return true;
    }
    Node<T> successor = getMinNode(curr.getRight());
    T k = successor.getX();
    remove(k);
    curr.setX(k);
    return true;
  }

  /**
   * searches for the first node with value x
   * @param x value of the node we are trying to find
   * @return TRUE if we found the node, and FALSE otherwise
   */
  public boolean find(T x){
    Node<T> curr = root();
    while (curr != null && curr.getX() != x){
      if (x.compareTo(curr.getX()) < 0) curr = curr.getLeft();
      else curr = curr.getRight();
    }
    return (curr != null && curr.getX() == x);
  }

  //---------------- TREE TRAVERSAL ------------------//

  /**
   * @return List of all the nodes in tree traversed In-Order
   */
  public List<Node<T>> inOrder() {
    List<Node<T>> v = new ArrayList<Node<T>>();
    if (!isEmpty()) inOrder(root(), v);
    return v;
  }

  /**
   * Fills the list with the nodes of the tree in an In-Order manner
   * @param p current node
   * @param v list that we are filling
   */
  private void inOrder(Node<T> p, List<Node<T>> v){
    if (p.getLeft() != null) inOrder(p.getLeft(), v);
    v.add(p);
    if (p.getRight() != null) inOrder(p.getRight(), v);
  }

  /**
   * @return List of all the nodes in tree traversed Pre-Order
   */
  public List<Node<T>> preOrder() {
    List<Node<T>> v = new ArrayList<Node<T>>();
    if (!isEmpty()) dfs(root(), v);
    return v;
  }

  /**
   * Fills the list with the nodes of the tree in an Pre-Order manner
   * @param p current node
   * @param v list that we are filling
   */
  private void dfs(Node<T> p, List<Node<T>> v){
    v.add(p);
    if (p.getLeft() != null) dfs(p.getLeft(), v);
    if (p.getRight() != null) dfs(p.getRight(), v);
  }

  /**
   * @return List of all the nodes in tree traversed Post-Order
   */
  public List<Node<T>> postOrder() {
    List<Node<T>> v = new ArrayList<Node<T>>();
    if (!isEmpty()) postOrder(root(), v);
    return v;
  }

  /**
   * Fills the list with the nodes of the tree in an Post-Order manner
   * @param p current node
   * @param v list that we are filling
   */
  private void postOrder(Node<T> p, List<Node<T>> v){
    if (p.getLeft() != null) postOrder(p.getLeft(), v);
    if (p.getRight() != null) postOrder(p.getRight(), v);
    v.add(p);
  }

  /**
   * Traverses the tree from top to bottom, from left to right
   * @return List of the all the nodes in the above specified order
   */
  public List<Node<T>> bfs() {
    List<Node<T>> v = new ArrayList<Node<T>>();
    if (isEmpty()) return v;
    Queue<Node<T>> q = new LinkedList<>();
    q.add(root);
    while (!q.isEmpty()){
      Node<T> temp = q.remove();
      if (temp.getLeft() != null) q.add(temp.getLeft());
      if (temp.getRight() != null) q.add(temp.getRight());
      v.add(temp);
    }
    return v;
  }

  /**
   * prints the nodes of the tree in a sideways manner
   */
  public void printSideways() {printSideways(root(), "");}
  /**
   * helper function to the above "printSideways" function
   * @param p current node
   * @param indent current indentation level
   */
  private void printSideways(Node<T> p, String indent){
    if (p == null) return;
    printSideways(p.getRight(), indent + "  ");
    System.out.println(indent + p.getX());
    printSideways(p.getLeft(), indent + "  ");
  }
}

public class Main {
  public static void main(String[] args){
    BinarySearchTree<Integer> tree = new BinarySearchTree<Integer>();
    Random rand = new Random();

    int sz = 20;
    for (int i = 0; i < sz; ++i) tree.add(rand.nextInt(100));

    tree.printSideways();
  }
}
